bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/handlers/AbstractSolaxHandler.java

```diff
@@ -0,0 +1,137 @@
+ */
+@NonNullByDefault
+public abstract class AbstractSolaxHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(AbstractSolaxHandler.class);
+
+    private static final int INITIAL_SCHEDULE_DELAY_SECONDS = 5;
+
+    private @NonNullByDefault({}) SolaxConnector connector;
+
+    private @Nullable ScheduledFuture<?> schedule;
+
+    private final ReentrantLock retrieveDataCallLock = new ReentrantLock();
+
+    protected final TranslationProvider i18nProvider;
+
+    protected final ZoneId timeZone;
+
+    public AbstractSolaxHandler(Thing thing, TranslationProvider i18nProvider, TimeZoneProvider timeZoneProvider) {
+        super(thing);
+        this.i18nProvider = i18nProvider;
+        this.timeZone = timeZoneProvider.getTimeZone();
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+
+        SolaxConfiguration config = getConfigAs(SolaxConfiguration.class);
+        connector = createConnector(config);
+        int refreshInterval = config.refreshInterval;
+        TimeUnit timeUnit = TimeUnit.SECONDS;
+
+        logger.debug("Scheduling regular interval retrieval every {} {}", refreshInterval, timeUnit);
+        schedule = scheduler.scheduleWithFixedDelay(this::retrieveData, INITIAL_SCHEDULE_DELAY_SECONDS, refreshInterval,
+                timeUnit);
+    }
+
+    private void retrieveData() {
+        if (retrieveDataCallLock.tryLock()) {
+            try {
+                String rawJsonData = connector.retrieveData();
+                logger.debug("Raw data retrieved = {}", rawJsonData);
+
+                if (rawJsonData != null && !rawJsonData.isEmpty()) {
+                    updateFromData(rawJsonData);
+                    if (getThing().getStatus() != ThingStatus.ONLINE) {
+                        updateStatus(ThingStatus.ONLINE);
+                    }
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            SolaxBindingConstants.I18N_KEY_OFFLINE_COMMUNICATION_ERROR_JSON_CANNOT_BE_RETRIEVED);
+                }
+            } catch (IOException e) {
+                logger.debug("Exception received while attempting to retrieve data via HTTP", e);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            } catch (SolaxUpdateException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR, e.getMessage());
+            } finally {
+                retrieveDataCallLock.unlock();
+            }
+        } else {
+            logger.debug("Unable to retrieve data because a request is already in progress.");
+        }
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            scheduler.execute(this::retrieveData);
```

Comment:
What is the idea here? To bring it offline forcefully with a proper status, so we don't end up in the API reject?

Reply:
Maybe if it is very close to the limit, bring ithe Thing offline until it is below some treshold (prevent flapping). That would mean you have to keep tracking of call timestamp. With 10k entry's that is maybe a but to expensive (performance wise).
It might get complex very fast. It's something to think about. The current solution is good enough to merge.